packages/psd/src/sections/LayerAndMaskInformation/AdditionalLayerInfo/readPatternAliBlock.ts

```diff
@@ -0,0 +1,189 @@
+  }
+
+  const scanlineSize = Array.from(Array(height), () =>
+    cursor.read("u16")
+  ).reduce((a, b) => a + b);
+  const data = cursor.take(scanlineSize);
+
+  return {
+    compression: ChannelCompression.RleCompressed,
+    data,
+  };
+}
+
+export function readChannel(
+  cursor: Cursor,
+  height: number
+): PatternDataChannel | null {
+  const written = cursor.read("u32");
+
+  if (!written) {
+    return null;
+  }
+
+  const length = cursor.read("u32");
+
+  if (!length) {
+    return null;
+  }
+
+  const {position} = cursor;
+
+  const pixelDepth1 = cursor.read("u32");
+  const rectangle = readRectangle(cursor);
+  const pixelDepth2 = cursor.read("u16");
+  const compression = cursor.read("u8");
+
+  const size = length - (cursor.position - position);
+  const channelBytes = readChannelBytes(cursor, size, compression, height);
+
+  return {
+    written: Boolean(written),
+    length,
+    pixelDepth1,
+    rectangle,
+    pixelDepth2,
+    ...channelBytes,
+  };
+}
+
+//https://www.adobe.com/devnet-apps/photoshop/fileformatashtml/#VirtualMemoryArrayList
+//The following is a virtual memory array, repeated for the number of channels + one for a user mask + one for a sheet mask.
+export function readPatternData(cursor: Cursor, height: number): PatternData {
+  const version = cursor.read("u32");
+  const length = cursor.read("u32");
+  const rectangle = readRectangle(cursor);
+  const numberOfChannels = cursor.read("u32");
+  const channels = new Map();
+
+  for (let i = 0; i < numberOfChannels + 2; i++) {
+    const channel = readChannel(cursor, height);
+    channels.set(i, channel);
+  }
+
+  return {version, length, rectangle, numberOfChannels, channels};
+}
+
+export function readPattern(cursor: Cursor): Pattern {
+  const version = cursor.read("u32");
+  const imageMode = cursor.read("u32");
+
+  const {vert: height, horiz: width} = readPoint(cursor);
+  const name = cursor.readUnicodeString(0);
+  const id = readPascalString(cursor);
+
+  const colorTable =
+    imageMode === ImageMode.Indexed ? readColorTable(cursor) : undefined;
+
+  const patternData = readPatternData(cursor, height);
+
+  return {
+    version,
+    imageMode,
+    width,
+    height,
+    name,
+    id,
+    ...(colorTable ? {colorTable} : null),
+    patternData,
+  };
+}
+
+export function readPatternAliBlock(
+  cursor: Cursor,
+  size: number
+): AliBlockBody<PatternAliBlock> {
+  if (size === 0) {
+    return {data: []};
+  }
+
+  const startPosition = cursor.position;
+
+  let currentPosition = cursor.position;
+  const data: Pattern[] = [];
+  while (currentPosition - startPosition < size) {
+    const patternLength = cursor.read("u32");
+    currentPosition = currentPosition + patternLength;
+
+    if (currentPosition - startPosition >= size) {
+      break;
+    }
+    data.push(readPattern(cursor));
+  }
+  cursor.pass(size - (currentPosition - startPosition));
+
+  return {
+    data,
+  };
```

Comment:
Maybe we can simplify this implementation a little?
```suggest
  const endAt = cursor.position + size;
  const data: Pattern[] = []

  let patternLength: number
  while (((cursor.position + 4) < endAt) && (patternLength = cursor.read("u32"))) {
    data.push(readPattern(cursor));
  }
  cursor.pass(endAt - cursor.position)
  return {data}
```

Reply:
great, just removed patternLength as it is not necessary
```
export function readPatternAliBlock(
  cursor: Cursor,
  size: number
): AliBlockBody<PatternAliBlock> {
  const endAt = cursor.position + size;
  const data: Pattern[] = [];

  while (cursor.position + 4 < endAt && cursor.read("u32")) {
    data.push(readPattern(cursor));
  }

  cursor.pass(endAt - cursor.position);
  return {data};
}
```